File: codegen/results/v002/apple-foundationmodel/20260411_223708/go/007_beatles_interview/solution.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

type Person struct {
	FirstName    string
	LastName     string
	Birthday     string
	Age          int
	Relatives    []map[string]string
	IsStillAlive bool
}

func parsePerson(row io.Reader, idx int) (*Person, error) {
	reader := bufio.NewReader(row)
	var firstName, lastName string
	var birthday, age int
	var relatives []map[string]string

	err := reader.ReadString('\n')
	if err != nil {
		return nil, fmt.Errorf("failed to read line: %w", err)
	}

	err = strings.Split(strings.TrimSpace(err), ",")
	if err != nil {
		return nil, fmt.Errorf("failed to split comma-separated values: %w", err)
	}

	firstName, err = strconv.Atoi(err[0])
	if err != nil {
		return nil, fmt.Errorf("failed to parse age: %w", err)
	}
	lastName, err = strconv.Atoi(err[1])
	if err != nil {
		return nil, fmt.Errorf("failed to parse age: %w", err)
	}
	birthday, err = strconv.Atoi(err[2])
	if err != nil {
		return nil, fmt.Errorf("failed to parse age: %w", err)
	}
	age = firstName * 100 + lastName
	relatives = make([]map[string]string, 0)

	var relativeFirstName, relativeLastName, relationship string
	var relationshipIndex int
	for idx < len(err) {
		if err[idx] == ',' {
			relationshipIndex++
			continue
		}
		if relationshipIndex > 0 {
			relativeFirstName, err = strconv.Atoi(err[idx])
			if err != nil {
				return nil, fmt.Errorf("failed to parse age: %w", err)
			}
			relativeLastName, err = strconv.Atoi(err[idx+1])
			if err != nil {
				return nil, fmt.Errorf("failed to parse age: %w", err)
			}
			relationship = err[idx+2]
			relatives = append(relatives, map[string]string{
				"FirstName": relativeFirstName,
				"LastName": relativeLastName,
				"Relationship": relationship,
			})
			idx++
		} else {
			relativeFirstName, err = strconv.Atoi(err[idx])
			if err != nil {
				return nil, fmt.Errorf("failed to parse age: %w", err)
			}
			relativeLastName, err = strconv.Atoi(err[idx+1])
			if err != nil {
				return nil, fmt.Errorf("failed to parse age: %w", err)
			}
			relationship = err[idx+2]
			relatives = append(relatives, map[string]string{
				"FirstName": relativeFirstName,
				"LastName": relativeLastName,
				"Relationship": relationship,
			})
			idx++
		}
	}

	return &Person{
		FirstName:    firstName,
		LastName:     lastName,
		Birthday:     birthday,
		Age:          age,
		Relatives:    relatives,
		IsStillAlive: false,
	}, nil
}

func calculateAge(dob string, baseDate time.Time) int {
	dobTime, err := time.Parse("02/01/1900", dob)
	if err != nil {
		return 0
	}
	return int(time.Since(baseDate).Seconds()) / 365
}

func main() {
	if len(os.Args) < 2 || os.Args[1] != "csv" {
		log.Fatal("expected input file to be a CSV")
	}

	filePath := fmt.Sprintf("input/%s", os.Args[1])
	if _, err := os.Stat(filePath); err != nil {
		log.Fatalf("failed to stat file: %s", filePath)
	}

	// Read CSV file
	file, err := os.Open(filePath)
	if err != nil {
		log.Fatalf("failed to open file: %s", filePath)
	}

	defer file.Close()

	var people []Person
	var idx int
	for {
		row, err := file.ReadString('\n')
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatalf("failed to read line: %s", err)
		}
		person, err := parsePerson(strings.NewReader(row), idx)
		if err != nil {
			log.Fatalf("failed to parse person: %s", err)
		}
		people = append(people, person)
		idx++
	}

	// Output JSON
	output, err := json.MarshalIndent(people, "", "  ")
	if err != nil {
		log.Fatalf("failed to marshal JSON: %s", err)
	}

	fmt.Println(string(output))
}

File: codegen/results/v002/apple-foundationmodel/20260411_223708/typescript/007_beatles_interview/solution.ts
const fs = require('fs');
const path = require('path');

function calculateAge(birthday: string, currentDate: Date) {
    const dateFormatter = new Date(birthday);
    const ageComponents = currentDate.getFullYear() - dateFormatter.getFullYear();
    const monthDiff = currentDate.getMonth() - dateFormatter.getMonth();
    const dayDiff = currentDate.getDate() - dateFormatter.getDate();
    
    if (monthDiff < 0 || (monthDiff === 0 && dayDiff < 0)) {
        ageComponents--;
    }
    
    return ageComponents;
}

async function main() {
    const inputFilePath = path.join(__dirname, 'input/input.csv');
    const expectedFormatFilePath = path.join(__dirname, 'input/expected_format.json');

    // Read the CSV file and parse it into an array of objects
    const csvData = fs.readFileSync(inputFilePath, 'utf8').toString().trim().split('\n');
    const csvRows = csvData.map(row => row.split(',').map(cell => cell.trim()));
    
    // Ensure the CSV file has the expected columns
    const expectedColumns = ['Name', 'Birthday', 'Died', 'Father', 'Mother', 'Brother', 'Sister'];
    if (!expectedColumns.every(column => csvRows[0].includes(column))) {
        throw new Error('CSV file does not have the expected columns.');
    }

    const people: any[] = [];

    csvRows.forEach(row => {
        const person: any = {
            FirstName: row[0],
            LastName: row[1],
            Birthday: row[2],
            Age: calculateAge(row[2], new Date()),
            Relatives: []
        };

        if (row.length > 2) {
            const father = row[3] === 'null' ? null : row[3];
            const mother = row[4] === 'null' ? null : row[4];
            const brother = row[5] === 'null' ? null : row[5];
            const sister = row[6] === 'null' ? null : row[6];

            person.Relatives.push({
                FirstName: brother || '',
                LastName: brother || '',
                Relationship: brother || '',
                ...(mother && { Relationship: 'Mother' }),
                ...(father && { Relationship: 'Father' }),
                ...(sister && { Relationship: 'Sister' })
            });
        }

        people.push(person);
    });

    // Convert people into a JSON array
    const jsonArray = JSON.stringify(people, null, 2);
    console.log(jsonArray);
}

main();